README.md
# Decentraland documentation

## Setup

1.  Add your site and author details in `_config.yml`.
2.  Add your Google Analytics, Disqus and MailChimp keys to `_config.yml`.
3.  Get a workflow going to see your site's output (with [CloudCannon](https://app.cloudcannon.com/) or Jekyll locally).

## Develop

Base was built with [Jekyll](http://jekyllrb.com/) version 3.4.3, but should support newer versions as well.

Install the dependencies with [Bundler](http://bundler.io/):

```bash
$ bundle install
```

Run `jekyll` commands through Bundler to ensure you're using the right versions:

```bash
$ bundle exec jekyll serve --incremental
```

## Editing

Please use Visual Studio Code with Prettier extension to have a consistent coding style

#### Posts

- Add, update or remove a post in the _Posts_ collection.
- The tutorials page is organised by categories.
- Change the defaults when new posts are created in `_posts/_defaults.md`.

#### Post Series

To create a new series:

- Add a new document to the `sets` collection.
- Set the `title` and `description`.

To add a tutorial/post to a series:

- Add a `set` field to the tutorial front matter which points to the file name of the desired set without the `.md` extention. e.g. If I have a set at `_sets/getting-started.md` I would use this in my tutorial front matter: `set: getting-started`.
- Add a `set_order` field to the tutorial front matter and specify a number. This is the tutorials order in the set.

#### Navigation

- Exposed as a data file to give clients better access.
- Set in the _Data_ / _Navigation_ section.

#### Footer

- Exposed as a data file to give clients better access.
- Set in the _Data_ / _Footer_ section.

---

Base template was made by [CloudCannon](http://cloudcannon.com/), the Cloud CMS for Jekyll.


_sets/releases.md
---
title: Release Notes
description: Release notes, highlighting the latest changes, updates, and warnings.
color: "#0093e3"
accent_color: "#0093e3"
order: 30
img: releases.svg
---


js/main.js
jQuery(function() {
  var states = {
    normal: 0,
    bottom: 1,
    top: 2
  };
  var latestState = states.normal;
  var $sidebar = $(".sidebar"),
    $content = $(".wrapper"),
    $tutorial = $(".main-container"),
    $header = $("header"),
    $postTopBar = $(".tutorial .title-search-wrapper");
  $postTopBarPosition = $postTopBar.offset();
  console.log($postTopBarPosition);
  offset = $content.offset().top + 60;

  var setActiveSidebarLink = throttle(function setActiveSidebarLink() {
    var $closest = getClosestHeader();
    if (!$closest.hasClass("active")) {
      $(".sidebar .current-post-list a").removeClass("active");
      $closest.addClass("active");
    }
  }, 100);

  var found = true;

  var $el;

  var href = $sidebar
    .find(".current-post-list a")
    .first()
    .attr("href");

  if (href !== undefined) {
    $(window).on("scroll", function() {
      setSidebar();
    });
  }

  function setSidebar() {
    var headerHeight = 48; // header of the whole page, with logo
    var offset = headerHeight;
    // $header.toggleClass("overflow", window.scrollY > headerHeight);

    var bottom = $tutorial.offset().top + $tutorial.outerHeight() - $sidebar.outerHeight() - offset;

    if (window.scrollY > bottom) {
      if (latestState !== states.bottom) {
        // sticky at the bottom
        $sidebar.css("position", "absolute").css("top", $tutorial.outerHeight() - $sidebar.outerHeight());
        latestState = states.bottom;
      }
    } else if (window.scrollY > $tutorial.offset().top - offset) {
      if (latestState !== states.top) {
        // sticky at the top
        $sidebar.css("position", "fixed").css("top", offset);
        latestState = states.top;
      }
    } else {
      if (latestState !== states.normal) {
        // normal no sticky
        $sidebar.css("position", "absolute").css("top", 0);
        latestState = states.normal;
      }
    }
  }

    $('.toggle-item').click(function(ev) {
        ev.preventDefault();
        var data = $(ev.target).attr('data-toggle')
        $('.toggle[data-id!="' + data + '"]').hide('fast')
        $('.toggle[data-id="' + data + '"]').toggle('fast')
    })
});

function getClosestHeader() {
  var $links = $(".internal-href"),
    top = window.scrollY,
    $last = $links.first(),
    $content = $(".tutorial-content");

  // console.log(top);

  if (top < 300) {
    return $last;
  }

  if (top + window.innerHeight >= $content.offset().top + $content.height()) {
    return $links.last();
  }

  for (var i = 0; i < $links.length; i++) {
    var $link = $links.eq(i),
      href = $link.attr("href");

    if (href !== undefined && href.charAt(0) === "#" && href.length > 1) {
      var $anchor = $(href);

      if ($anchor.length > 0) {
        var offset = $anchor.offset();

        if (top < offset.top - window.innerHeight / 2) {
          return $last;
        }

        $last = $link;
      }
    }
  }
  return $last;
}

function throttle(callback, limit) {
  var wait = false;
  return function() {
    if (!wait) {
      callback.apply(null, arguments);
      wait = true;
      setTimeout(function() {
        wait = false;
      }, limit);
    }
  };
}


js/header.js
---
layout: null
---
$(function() {
  function closeSidebar() {
    $(".sidebar").removeClass("open")
  }

  function openDropdown() {
    closeSearchResults()
    closeSidebar()
    $("header .dropdown-trigger").addClass("open")
    $("header .dropdown-content").addClass("open")
    $("header .dropdown-overlay").addClass("open")
  }

  function closeDropdown() {
    $("header .dropdown-trigger").removeClass("open")
    $("header .dropdown-content").removeClass("open")
    $("header .dropdown-overlay").removeClass("open")
  }

  function openSearchResults() {
    $(".header_search").addClass("open")
    $(".header_search .search-results").empty()
    $("header .search-overlay").addClass("open")
  }

  function closeSearchResults() {
    $(".header_search").removeClass("open")
    $("header .search-overlay").removeClass("open")
  }

  $("header .dropdown-trigger").click(function(event) {
    event.preventDefault()
    $(this).hasClass("open") ? closeDropdown() : openDropdown()
  })

  $("header .dropdown-overlay").click(function(event) {
    event.preventDefault()
    closeDropdown()
  })

  $("header .search-overlay").click(function(event) {
    event.preventDefault()
    closeSearchResults()
  })

  const $searchInput = $('.header_search input[type="search"]')

  $("header .close").click(function(event) {
    event.preventDefault()
    closeSearchResults()
    $searchInput.val('')
  })

  $searchInput.focus(function() {
    closeSidebar()
    closeDropdown()
    showSearchResults()
  })

  function zoomDisable() {
    $('head meta[name=viewport]').remove()
    $('head').prepend('<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=0" />')
  }

  function zoomEnable() {
    $('head meta[name=viewport]').remove()
    $('head').prepend('<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=1" />')
  }

  const $inputs = $('.header_search input[type="search"], select, .newsletter input[type="email"]')

  $inputs.on('touchstart', function() {
    zoomDisable()
  })

  $inputs.on('touchend', function() {
    setTimeout(zoomEnable, 500)
  })

  function getPreview(query, content, previewLength) {
    previewLength = previewLength || content.length * 2

    var parts = query.split(" "),
      match = content.toLowerCase().indexOf(query.toLowerCase()),
      matchLength = query.length,
      preview

    // Find a relevant location in content
    for (var i = 0; i < parts.length; i++) {
      if (match >= 0) {
        break
      }

      match = content.toLowerCase().indexOf(parts[i].toLowerCase())
      matchLength = parts[i].length
    }

    // Create preview
    if (match >= 0) {
      var start = match - previewLength / 2,
        end = start > 0 ? match + matchLength + previewLength / 2 : previewLength

      preview = content.substring(start, end).trim()

      if (start > 0) {
        preview = "..." + preview
      }

      if (end < content.length) {
        preview = preview + "..."
      }

      // Highlight query parts
      preview = preview.replace(
        new RegExp("(" + parts.join("|") + ")", "gi"),
        "<strong>$1</strong>"
      )
    } else {
      // Use start of content if no match found
      preview =
        content.substring(0, previewLength).trim() +
        (content.length > previewLength ? "..." : "")
    }

    return preview
  }

  function showSearchResults() {
    const userInput = $searchInput.val().toLowerCase()

    if (userInput.length === 0) {
      closeSearchResults()
      return
    }

    openSearchResults()

    const items = window.index
      .search(userInput)
      .map(index => window.data[index.ref])

    const limit = 4
    const results = items.slice(0, limit)
    const $list = $('.header_search .search-results')

    for (result of results) {
      let category = result.categories.split(',')[0]
      if (category === 'Decentraland') {
        category = 'home'
      }

      $list.append(
        `<li>
          <a href="${result.url}">
            <div class="icon">
              <img src="{{ site.baseurl }}/images/sets/${category}.svg" />
            </div>
            <div>
              <span class="title">${result.title}</span>
              <span class="description">${getPreview(userInput, result.content, 120)}</span>
            </div>
          </a>
        </li>`
      )
    }

    if (items.length > limit) {
      $list.append(
        `<li class="more-results">
          <a href="/search/?q=${userInput}">See more results</a>
        </li>`
      )
    }

    $list.find('li')
      .mouseenter(function() {
        $list.find('li.selected').removeClass('selected')
        $(this).addClass('selected')
      })
      .mouseleave(function() {
        $(this).removeClass('selected')
      })
  }

  $searchInput.keydown(function(event) {
    const $list = $('.header_search .search-results')
    const $selected = $list.find('li.selected')

    let $next

    function selectNextItem() {
      event.preventDefault()
      $selected.removeClass('selected')
      $next.addClass('selected')
    }

    switch (event.key) {
      case 'Down': // IE specific value
      case 'ArrowDown':
        $next = $selected.next()
        if ($next.length === 0) {
          $next = $list.find('li:first-child')
        }
        selectNextItem()
        break

      case 'Up': // IE specific value
      case 'ArrowUp':
        $next = $selected.prev()
        if ($next.length === 0) {
          $next = $list.find('li:last-child')
        }
        selectNextItem()
        break

      case 'Enter':
        if ($selected.length > 0) {
          event.preventDefault()
          document.location.href = $selected.find('a').attr('href')
        }
        break
    }
  })

  let fetching = false

  $searchInput.on('input', function() {
    if (fetching) return

    if (window.data) {
      showSearchResults()
      return
    }

    fetching = true

    $.getJSON('{{ site.baseurl }}/data.json', function(data) {
      fetching = false
      window.data = data
      window.index = lunr(function() {
        this.field('id')
        this.field('title', { boost: 10 })
        this.field('categories')
        this.field('url')
        this.field('content')
      })

      for (var key in window.data) {
        window.index.add(window.data[key])
      }

      showSearchResults()
    })
  })
})


js/search.js
---
layout: null
---
(function () {
	function getQueryVariable(variable) {
		var query = window.location.search.substring(1),
			vars = query.split("&");

		for (var i = 0; i < vars.length; i++) {
			var pair = vars[i].split("=");

			if (pair[0] === variable) {
				return decodeURIComponent(pair[1].replace(/\+/g, '%20')).trim();
			}
		}
	}

	function displaySearchResults(results, query) {
		var searchResultsEl = document.getElementById("search-results"),
			searchProcessEl = document.getElementById("search-process");

		if (results.length) {
			var resultsHTML = "";
			results.forEach(function (result) {
				var item = window.data[result.ref],
					contentPreview = getPreview(query, item.content, 170),
					titlePreview = getPreview(query, item.title);

				let category = item.categories.split(',')[0]
				if (category === 'Decentraland') {
					category = 'general'
				}

				resultsHTML += '<li>' +
					'<a href="{{ site.baseurl }}' + item.url.trim() + '">' +
						'<div>' +
							'<div class="icon">' +
								'<img src="{{ site.baseurl }}/images/sets/' + category + '.svg" />' +
							'</div>' +
							'<div>' +
								'<h4>' + titlePreview + '</h4>' +
								'<p>' + contentPreview + '</p>' +
							'</div>' +
						'</div>' +
					'</a>' +
				'</li>';
			});

			searchResultsEl.innerHTML = resultsHTML;
			searchProcessEl.innerText = "Search";
		} else {
			searchResultsEl.style.display = "none";
			searchProcessEl.innerText = "No";
		}
	}

	window.index = lunr(function () {
		this.field("id");
		this.field("title", {boost: 10});
		this.field("categories");
		this.field("url");
		this.field("content");
	});

	var query = decodeURIComponent((getQueryVariable("q") || "").replace(/\+/g, "%20")),
		searchQueryContainerEl = document.getElementById("search-query-container"),
		searchQueryEl = document.getElementById("search-query");

	searchQueryEl.innerText = query;
	searchQueryContainerEl.style.display = "inline";

	for (var key in window.data) {
		window.index.add(window.data[key]);
	}

	displaySearchResults(window.index.search(query), query); // Hand the results off to be displayed
})();

js/global.js
---
layout: null
---

function indexData(data) {
  var index = lunr(function() {
    this.field('id')
    this.field('title', { boost: 10 })
    this.field('categories')
    this.field('url')
    this.field('content')
  })

  for (var key in data) {
    index.add(data[key])
  }

  window.index = index
  window.data = data
}

if (!document.location.pathname.match('/search')) {
  $.getJSON('{{ site.baseurl }}/data.json?{{ site.time | date: "%s%N" }}', function(data) {
    indexData(data)
  })
}

function getPreview(query, content, previewLength) {
  previewLength = previewLength || content.length * 2

  var parts = query.split(" "),
    match = content.toLowerCase().indexOf(query.toLowerCase()),
    matchLength = query.length,
    preview

  // Find a relevant location in content
  for (var i = 0; i < parts.length; i++) {
    if (match >= 0) {
      break
    }

    match = content.toLowerCase().indexOf(parts[i].toLowerCase())
    matchLength = parts[i].length
  }

  // Create preview
  if (match >= 0) {
    var start = match - previewLength / 2,
      end = start > 0 ? match + matchLength + previewLength / 2 : previewLength

    preview = content.substring(start, end).trim()

    if (start > 0) {
      preview = "..." + preview
    }

    if (end < content.length) {
      preview = preview + "..."
    }

    // Highlight query parts
    preview = preview.replace(
      new RegExp("(" + parts.join("|") + ")", "gi"),
      "<strong>$1</strong>"
    )
  } else {
    // Use start of content if no match found
    preview =
      content.substring(0, previewLength).trim() +
      (content.length > previewLength ? "..." : "")
  }

  return preview
}

$(function() {
  // HEADER ==>

  const $header = $('header')
  const $headerSearch = $('.header_search')

  function openDropdown() {
    closeSearchResults()
    closeSidebar()
    $header.find('.dropdown-trigger').addClass('open')
    $header.find('.dropdown-content').addClass('open')
    $header.find('.dropdown-overlay').addClass('open')
  }

  function closeDropdown() {
    $header.find('.dropdown-trigger').removeClass('open')
    $header.find('.dropdown-content').removeClass('open')
    $header.find('.dropdown-overlay').removeClass('open')
  }

  function openSearchResults() {
    $headerSearch.addClass('open')
    $headerSearch.find('.search-results').empty()
    $header.find('.search-overlay').addClass('open')
  }

  function closeSearchResults() {
    $headerSearch.removeClass('open')
    $header.find('.search-overlay').removeClass('open')
  }

  $header.find('.dropdown-trigger').click(function(event) {
    event.preventDefault()
    $(this).hasClass('open') ? closeDropdown() : openDropdown()
  })

  $header.find('.dropdown-overlay').click(function(event) {
    event.preventDefault()
    closeDropdown()
  })

  $header.find('.search-overlay').click(function(event) {
    event.preventDefault()
    closeSearchResults()
  })

  const $searchInput = $headerSearch.find('input[type="text"]')

  $header.find('.close').click(function(event) {
    event.preventDefault()
    closeSearchResults()
    $searchInput.val('')
  })

  $searchInput.on('focus click', function() {
    closeSidebar()
    closeDropdown()
    showSearchResults()
  })

  function zoomDisable() {
    $('head meta[name=viewport]').remove()
    $('head').prepend('<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=0" />')
  }

  function zoomEnable() {
    $('head meta[name=viewport]').remove()
    $('head').prepend('<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=1" />')
  }

  const touchableInputs = [
    '.header_search input[type="text"]',
    '.newsletter input[type="email"]',
    'select',
  ]
  const $inputs = $(touchableInputs.join(','))

  $inputs.on('touchstart', function() {
    zoomDisable()
  })

  $inputs.on('touchend', function() {
    setTimeout(zoomEnable, 500)
  })

  function showSearchResults() {
    const userInput = $searchInput.val().toLowerCase()

    if (userInput.length === 0) {
      closeSearchResults()
      return
    }

    openSearchResults()

    const items = window.index
      .search(userInput)
      .map(function (index) { return window.data[index.ref] })

    const limit = 4
    const results = items.slice(0, limit)
    const $list = $headerSearch.find('.search-results')

    for (var i = 0; i < results.length; i++) {
      var result = results[i]
      var category = result.categories.split(',')[0]

      if (category === 'Decentraland') {
        category = 'general'
      }

      $list.append(
        '<li>' +
          '<a href="' + result.url + '">' +
            '<div class="icon">' +
              '<img src="{{ site.baseurl }}/images/sets/' + category + '.svg" />' +
            '</div>' +
            '<div>' +
              '<span class="title">' + result.title + '</span>' +
              '<span class="description">' + getPreview(userInput, result.content, 120) + '</span>' +
            '</div>' +
          '</a>' +
        '</li>'
      )
    }

    if (items.length > limit) {
      $list.append(
        '<li class="more-results">' +
          '<a href="{{ site.baseurl }}/search/?q=' + userInput + '">See more results</a>' +
        '</li>'
      )
    }

    if (results.length === 0) {
      $list.append(
        '<li class="no-results">' +
          '<div class="image">' +
            '{% include search-big.svg %}' +
          '</div>' +
          '<strong>Sorry, we couldn\'t find any matches</strong>' +
          '<span>Try searching for a different keyword</span>' +
        '</li>'
      )
    }

    $list.find('li:not(.no-results)')
      .mouseenter(function() {
        $list.find('li.selected').removeClass('selected')
        $(this).addClass('selected')
      })
      .mouseleave(function() {
        $(this).removeClass('selected')
      })
  }

  $searchInput.keydown(function(event) {
    const $list = $headerSearch.find('.search-results')
    const $selected = $list.find('li.selected')

    let $next

    function selectNextItem() {
      event.preventDefault()
      $selected.removeClass('selected')
      $next.addClass('selected')
    }

    switch (event.key) {
      case 'Down': // IE specific value
      case 'ArrowDown':
        $next = $selected.next()
        if ($next.length === 0) {
          $next = $list.find('li:first-child')
        }
        selectNextItem()
        break

      case 'Up': // IE specific value
      case 'ArrowUp':
        $next = $selected.prev()
        if ($next.length === 0) {
          $next = $list.find('li:last-child')
        }
        selectNextItem()
        break

      case 'Escape':
        closeSearchResults()
        break

      case 'Enter':
        if ($selected.length > 0) {
          event.preventDefault()
          document.location.href = $selected.find('a').attr('href')
        }
        break
    }
  })

  let fetching = false

  $searchInput.on('input', function() {
    if (fetching) return

    if (window.data) {
      showSearchResults()
      return
    }

    fetching = true
    $headerSearch.addClass('fetching')

    $.getJSON('{{ site.baseurl }}/data.json?{{ site.time | date: "%s%N" }}', function(data) {
      fetching = false
      $headerSearch.removeClass('fetching')

      indexData(data)
      showSearchResults()
    })
  })

  // SIDEBAR ==>

  const $sidebar = $('.sidebar')
  const $sidebarDropdown = $sidebar.find('.dropdown')

  function closeSidebar() {
    $sidebar.removeClass('open')
  }

  $sidebarDropdown.click(function(event) {
    event.preventDefault()
    $sidebar.addClass('open')
    $('body').addClass('modal-open')
  })

  $sidebar.find('.close').click(function(event) {
    event.preventDefault()
    closeSidebar()
    $('body').removeClass('modal-open')
  })

  $sidebar.find('.toggle-item').click(function(event) {
    event.preventDefault()
    const data = $(event.target).attr('data-toggle')
    $('.toggle[data-id!="' + data + '"]').hide('fast')
    $('.toggle[data-id="' + data + '"]').toggle('fast')
  })

  const $window = $(window)
  $window.bind('scroll', function () {
    const $languageSelector = $('.select-language.visible')
    const offset = $header.height() + $header.offset().top
    const threshold = ($languageSelector.length > 0)
      ? offset + $languageSelector.height()
      : offset

    $sidebarDropdown.toggleClass('sticky', $window.scrollTop() > threshold)
  })

  // HEADINGS ==>

  const headings = document.querySelectorAll('h2[id]')

  for (var i = 0; i < headings.length; i++) {
    const anchorLink = document.createElement('a')
    anchorLink.innerText = '#'
    anchorLink.href = '#' + headings[i].id
    anchorLink.classList.add('header-link')

    headings[i].appendChild(anchorLink)
  }

  $('a[href*=\\#]').not('.no-smooth').on('click', function() {
    const $el = $(this.hash)
    if ($el.length > 0) {
      $('html,body').animate({ scrollTop: $el.offset().top - 30 }, 500)
    }
  })

  // LANGUAGE ==>

  function dismissLanguageBar() {
    Cookies.set('language', true, { expires: 365 })
    $('.select-language').removeClass('visible')
  }

  $('select#lang').on('change', function() {
    const value = $(this).val()
    dismissLanguageBar()
    $('select#lang').val(value)
    $('.lang-flag')
      .removeClass('english chinese')
      .addClass(value)

    const sites = {
      english: 'https://docs.decentraland.org',
      chinese: 'https://cn.docs.decentraland.org'
    }

    document.location.href = sites[value] + document.location.pathname
  })

  $('.select-language .dismiss').click(function() {
    dismissLanguageBar()
  })

  if (!Cookies.get('language')) {
    $('.select-language').addClass('visible')
  }

  // FEEDBACK ==>

  const $feedback = $('.feedback')
  const $textarea = $feedback.find('.textarea')
  const $input = $textarea.find('textarea')
  const $mirror = $textarea.find('.mirror')
  const $send = $feedback.find('.send')

  let articleWasUseful

  function sendingFeedback(value) {
    articleWasUseful = value

    resetFeedback()

    $feedback
      .removeClass('yes no')
      .addClass('sending ' + value)

    $textarea.click()
  }

  $feedback.find('.yes').click(function() {
    sendingFeedback('yes')
  })

  $feedback.find('.no').click(function() {
    sendingFeedback('no')
  })

  function resetFeedback() {
    $input.val('')
    $mirror.text('')
  }

  function sendFeedback(withFeedback) {
    const payload = {
      useful: articleWasUseful,
      feedback: withFeedback ? $input.val().trim() : null
    }

    {% if jekyll.environment == 'production' and site.segment_write_key != '' %}
      analytics.track('Article Feedback', payload)
    {% else %}
      console.log(payload)
    {% endif %}

    $send.prop({ disabled: true })
    $input.blur()
    $feedback
      .removeClass('sending')
      .addClass('sent')
  }

  $send.click(function() {
    sendFeedback(true)
  })

  $feedback.find('.skip').click(function() {
    sendFeedback(false)
  })

  $textarea.on('click', function() {
    zoomDisable()
    $input.focus()
  })

  $input.on('keydown keyup', function(event) {
    const value = event.target.value;
    $send.prop({ disabled: value.trim().length === 0 })
    $mirror.text(value)
  })

  $input.on('keydown', function(event) {
    switch (event.key) {
      case 'Enter':
        if (!$send.attr('disabled')) {
          sendFeedback(true)
        } else {
          event.preventDefault()
        }
        break
    }
  })

  // SCROLLABLE TABLES ==>

  $('.tutorial-main table').wrap('<div class="scrollable"></div>')
})
